Library/Homebrew/cmd/uninstall.rb

```diff
@@ -26,6 +53,36 @@ def uninstall
         name = Formulary.canonical_name(name)
         rack = HOMEBREW_CELLAR/name
 
+        # Cellar/oldname can be a symlink to Cellar/newname so in order not to
+        # get newname's keg as oldname's keg we check if it's a symlink.
+        if rack.symlink?
+          raise "#{rack.to_s} is a symlink"
+        end
```

Comment:
> because we want to be able to rename formulae and replace them with other things.

I don't think this is contrary to treating it alias. Because we only resolve the alias after no formula is found, that is if a new formula occupy this name the old name alias will be shadowed.